PythonVisualizations/SortingBase.py

```diff
@@ -0,0 +1,420 @@
+        upDelta = (0, - self.CELL_SIZE * 4 // 3)
+        downDelta = multiply_vector(upDelta, -1)
+        if a == b:  # Swapping with self - just move up & down
+            self.moveItemsBy(itemsA, upDelta, sleepTime=0.02)
+            self.moveItemsBy(itemsA, downDelta, sleepTime=0.02)
+            return
+    
+        # make a and b cells plus their associated items switch places
+        self.moveItemsOnCurve(
+            itemsA + itemsB, [self.canvas.coords(i) for i in itemsB + itemsA],
+            sleepTime=0.05, startAngle=90 * 11 / (10 + abs(a - b)))
+    
+        # perform the actual cell swap operation in the list
+        self.list[a], self.list[b] = self.list[b], self.list[a]   
+        
+    def createIndex(  # Create an index arrow to point at an indexed
+            self, index,  # cell
+            name=None,  # with an optional name label
+            level=1,  # at a particular level away from the cells
+            color=None):  # (negative are below)
+        if not color: color = self.VARIABLE_COLOR
+
+        cell_coords = self.cellCoords(index)
+        cell_center = self.cellCenter(index)
+        level_spacing = self.VARIABLE_FONT[1]
+        x = cell_center[0]
+        if level > 0:
+            y0 = cell_coords[1] - self.CELL_SIZE * 3 // 5 - level * level_spacing
+            y1 = cell_coords[1] - self.CELL_SIZE * 3 // 10
+        else:
+            y0 = cell_coords[3] + self.CELL_SIZE * 3 // 5 - level * level_spacing
+            y1 = cell_coords[3] + self.CELL_SIZE * 3 // 10
+        arrow = self.canvas.create_line(
+            x, y0, x, y1, arrow="last", fill=color)
+        if name:
+            label = self.canvas.create_text(
+                x + 2, y0, text=name, anchor=SW if level > 0 else NW,
+                font=self.VARIABLE_FONT, fill=color)
+        return (arrow, label) if name else (arrow,)  
+ 
+
+    # ARRAY FUNCTIONALITY
+    def insert(self, val):
+        self.startAnimations()
+        callEnviron = self.createCallEnvironment()
+        
+        # Inserting when there is no room on the canvas
+        if self.window.winfo_width() <= self.ARRAY_X0 + (
+            (len(self.list)+1) * self.CELL_SIZE):
+            # change cell width to 20
+            self.CELL_WIDTH = 20
+            # if this is the first insert that will change the width 
+            if not self.changeSize:
+                self.changeSize = True 
+                # redraw the existing cells to be narrower
+                self.fixCells()
+                self.redrawArrayCells()
+                # delete the cell values 
+                self.canvas.delete("cellVal")
+            
+        # If array needs to grow, add cells:
+        while self.size < len(self.list) + 1:
+            self.size += 1
+            self.createArrayCell(len(self.list))
+
+        # draw an index pointing to the last cell
+        indexDisplay = self.createIndex(len(self.list), 'nItems')
+        callEnviron |= set(indexDisplay)
```

Comment:
It might not matter now, but instead of `self.window.winfo_width()` on line 131, I would use `self.canvas.winfo_width()`.  I think they currently are equal, but if we ever put another widget to the right or left of the canvas, the width would no longer be correct.  Maybe the best thing to do would be to move Line 158 up to before Line 130 and use the values in `canvasDimensions`.

Shouldn't the calculation on line 132 use `self.CELL_WIDTH` instead of `self.CELL_SIZE`?  Even when using the smaller cells, we want to prevent users from creating cells that aren't visible. 

Thirdly, there is no `else` clause to the `if` on line 136 that checks if the smaller cells are being used.  Users should get an error message if they try to insert values that won't be visible.  We should probably also have an upper limit on the number of cells (that can be customized by the subclasses of `SortingBase`).   Even if there's enough room on the canvas, trying to insert beyond that limit would trigger an error (perhaps with a different error message such as 'Array is limited to NNN cells' versus 'Inserted value would not be visible').

I think the code is cleaner if we get rid of the `redrawArrayCells()` method and simply call `display()` to redisplay the array and `nItems` pointer.  Using `display()` as it is currently defined would erase any temporary canvas items.  That's probably the correct thing to do, but if there were some case where they should be preserved (e.g. insert being called from inside another operation), they probably would need to be moved due to the change in cell positions (and the caller would have to take of that).  This is part of the `insert()` method, and that only involves the `nItems` index.  So I think just using `display()` after changing the cell width is best.  That makes lines 133 - 151 collapse to:
```
            if not self.changeSize:
                self.changeSize = True 
                self.CELL_WIDTH = 20
                self.size = max(self.size, len(self.list) + 1)
                # redraw the existing cells to be narrower
                self.display()
```

Reply:
Fixed in commit 8169bd2 (and earlier).